### PR TITLE
Agent save history

### PR DESCRIPTION
**Pull request by yoptar**

### CODE REVIEW

File: deeppavlov/core/agent_v2/state_schema.py [hunk 1]
```diff
@@ -14,9 +14,14 @@ def to_dict(self):
                 'personality': self.personality}
 
 
+class Bot(User):
+    user_type = StringField(required=True, choices=['bot'], default='bot')
```
Comment (my-master) [inline deeppavlov/core/agent_v2/state_schema.py]: no need in `choices`

File: deeppavlov/core/agent_v2/state_schema.py [hunk 2]
```diff
@@ -14,9 +14,14 @@ def to_dict(self):
...
+
+
 class Human(User):
+    user_type = StringField(required=True, choices=['human'], default='human')
```
Comment (my-master) [inline deeppavlov/core/agent_v2/state_schema.py]: no need in `choices`

File: deeppavlov/core/agent_v2/state_schema.py (hunk 1 above)
Comment (my-master) [reply deeppavlov/core/agent_v2/state_schema.py]: `to_dict` should be moved from `User` to `Bot`, if we create a separate `Bot` class.

File: deeppavlov/core/agent_v2/state_schema.py [hunk 3]
```diff
@@ -72,15 +77,16 @@ def to_dict(self):
...
         return {
             'id': str(self.id),
             'location': self.location,
             'utterances': [utt.to_dict() for utt in self.utterances],
-            'user': [u.to_dict() for u in self.users if u.user_type == 'human'][0],
-            'bot': [u.to_dict() for u in self.users if u.user_type == 'bot'][0],
+            'user': self.user.to_dict(),
+            'bot': self.bot.to_dict(),
             'channel_type': self.channel_type
         }
```
Comment (my-master) [inline deeppavlov/core/agent_v2/state_schema.py]: 👍

File: deeppavlov/core/agent_v2/state_manager.py [hunk 4]
```diff
@@ -10,49 +10,71 @@
...
         utterances = []
         if annotations is None:
             annotations = [None] * len(texts)
         for text, anno, user, date_time in zip(texts, annotations, users, date_times):
-            utterances.append(self.create_new_utterance(text, user, date_time, anno))
+            utterances.append(cls.create_new_utterance(text, user, date_time, anno))
+        return utterances
+
+    @classmethod
+    def add_bot_utterances(cls, dialogs: Collection[Dialog], texts, date_times, active_skills, confidences,
```
[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/core/agent_v2/state_manager.py]: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes
`Sequence` is broader then `Collection`

File: deeppavlov/core/agent_v2/state_schema.py [hunk 5]
```diff
@@ -14,9 +14,14 @@ def to_dict(self):
...
 
+class Bot(User):
+    user_type = StringField(required=True, choices=['bot'], default='bot')
+
+
 class Human(User):
+    user_type = StringField(required=True, choices=['human'], default='human')
     user_telegram_id = StringField(required=True, unique=True, sparse=True)
-    device_type = DynamicField()
+    # device_type = DynamicField()
```
[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/core/agent_v2/state_schema.py]: :heavy_check_mark:

File: deeppavlov/core/agent_v2/state_schema.py (hunk 1 above)
Comment (yoptar) [reply deeppavlov/core/agent_v2/state_schema.py]: Removed the field altogether

File: deeppavlov/core/agent_v2/state_schema.py (hunk 2 above)
Comment (yoptar) [reply deeppavlov/core/agent_v2/state_schema.py]: Removed the field altogether

File: deeppavlov/core/agent_v2/state_manager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (my-master) [reply deeppavlov/core/agent_v2/state_manager.py]: It's not broader, it inherits from `Collection` and `Reversible`, so has more methods than `Collection`, so is narrower.
https://data-flair.training/blogs/python-sequence/

[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/core/agent_v2/state_manager.py]: You're right

**Pull request merged**